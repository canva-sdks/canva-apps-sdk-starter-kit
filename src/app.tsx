import { Button, Rows, Text } from "@canva/app-ui-kit";
import { FormattedMessage, useIntl } from "react-intl";
import * as styles from "styles/components.css";
import { useAddElement } from "utils/use_add_element";

export const App = () => {
  const addElement = useAddElement();
  const onClick = () => {
    addElement({
      type: "text",
      children: ["Hello world!"],
    });
  };

  const intl = useIntl();

  return (
    <div className={styles.scrollContainer}>
      <Rows spacing="2u">
        <Text>
          <FormattedMessage
            defaultMessage="
              To make changes to this app, edit the <code>src/app.tsx</code> file,
              then close and reopen the app in the editor to preview the changes.
            "
            description="Instructions for how to make changes to the app. Do not translate <code>src/app.tsx</code>."
            values={{
              code: (chunks) => <code>{chunks}</code>,
            }}
          />
        </Text>
        <Button variant="primary" onClick={onClick} stretch>
          {intl.formatMessage({
            defaultMessage: "Do something cool",
            description:
              "Button text to do something cool. Creates a new text element when pressed.",
          })}
        </Button>
      </Rows>
    </div>
  );
};
