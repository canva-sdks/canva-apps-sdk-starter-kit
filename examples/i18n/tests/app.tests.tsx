import type { RenderResult } from "@testing-library/react";
import { fireEvent, render } from "@testing-library/react";
import { TestAppUiProvider } from "@canva/app-ui-kit";
import { TestAppI18nProvider } from "@canva/app-i18n-kit";
import { App, TokenCount } from "../app";

function renderInTestProvider(node: React.ReactNode): RenderResult {
  return render(
    // In a test environment, you should wrap your apps in `TestAppI18nProvider` and `TestAppUiProvider`, rather than `AppI18nProvider` and `AppUiProvider`
    <TestAppI18nProvider>
      <TestAppUiProvider>{node}</TestAppUiProvider>
    </TestAppI18nProvider>,
  );
}

describe("app", () => {
  let requestOpenExternalUrl: jest.Mock;

  beforeEach(() => {
    requestOpenExternalUrl = jest.fn().mockResolvedValue({});
    jest.useFakeTimers({
      now: new Date("2024-09-25"), // For consistent snapshots, pretend today is always Canva Extend 2024
    });
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it("calls openExternalUrl onClick", async () => {
    const result = renderInTestProvider(
      <App requestOpenExternalUrl={requestOpenExternalUrl} />,
    );

    // get a reference to the link element
    const galleryExternalLink = result.getByText(/gallery/);

    // assert its label matches what we expect
    expect(galleryExternalLink.textContent).toContain("gallery");
    // assert our callback has not yet been called
    expect(requestOpenExternalUrl).not.toHaveBeenCalled();

    // programmatically simulate clicking the button
    fireEvent.click(galleryExternalLink);

    // assert our callback was called
    expect(requestOpenExternalUrl).toHaveBeenCalledTimes(1);
  });

  it("Renders token counts consistently ðŸŽ‰", () => {
    const resultToken0 = renderInTestProvider(
      <TokenCount used={0} total={25} />,
    );
    // The snapshot test can be used to detect unexpected changes in the rendered output.
    expect(resultToken0.container).toMatchSnapshot();

    const resultToken1 = renderInTestProvider(
      <TokenCount used={1} total={25} />,
    );
    expect(resultToken1.container).toMatchSnapshot();

    const resultToken10 = renderInTestProvider(
      <TokenCount used={10} total={25} />,
    );
    expect(resultToken10.container).toMatchSnapshot();
  });
});
