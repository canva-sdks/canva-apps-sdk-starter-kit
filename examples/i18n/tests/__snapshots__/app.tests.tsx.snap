// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app Renders token counts consistently 🎉 1`] = `
<div>
  <p
    class="fFOiLQ _5Ob_nQ fM_HdA"
  >
    You've used 0 tokens out of 25.
  </p>
</div>
`;

exports[`app Renders token counts consistently 🎉 2`] = `
<div>
  <p
    class="fFOiLQ _5Ob_nQ fM_HdA"
  >
    You've used 1 token out of 25.
  </p>
</div>
`;

exports[`app Renders token counts consistently 🎉 3`] = `
<div>
  <p
    class="fFOiLQ _5Ob_nQ fM_HdA"
  >
    You've used 10 tokens out of 25.
  </p>
</div>
`;
